crates/egui/src/containers/popup.rs

```diff
@@ -1,457 +1,517 @@
+    Position(Pos2),
+}
+
+impl From<Rect> for PopupAnchor {
+    fn from(rect: Rect) -> Self {
+        Self::Rect(rect)
     }
 }
 
-fn remember_that_tooltip_was_shown(ctx: &Context) {
-    let now = ctx.input(|i| i.time);
-    ctx.data_mut(|data| data.insert_temp::<f64>(when_was_a_toolip_last_shown_id(), now));
+impl From<Pos2> for PopupAnchor {
+    fn from(pos: Pos2) -> Self {
+        Self::Position(pos)
+    }
 }
 
-// ----------------------------------------------------------------------------
-
-/// Show a tooltip at the current pointer position (if any).
-///
-/// Most of the time it is easier to use [`Response::on_hover_ui`].
-///
-/// See also [`show_tooltip_text`].
-///
-/// Returns `None` if the tooltip could not be placed.
-///
-/// ```
-/// # egui::__run_test_ui(|ui| {
-/// if ui.ui_contains_pointer() {
-///     egui::show_tooltip(ui.ctx(), ui.layer_id(), egui::Id::new("my_tooltip"), |ui| {
-///         ui.label("Helpful text");
-///     });
-/// }
-/// # });
-/// ```
-pub fn show_tooltip<R>(
-    ctx: &Context,
-    parent_layer: LayerId,
-    widget_id: Id,
-    add_contents: impl FnOnce(&mut Ui) -> R,
-) -> Option<R> {
-    show_tooltip_at_pointer(ctx, parent_layer, widget_id, add_contents)
+impl From<&Response> for PopupAnchor {
+    fn from(response: &Response) -> Self {
+        let mut widget_rect = response.rect;
+        if let Some(to_global) = response.ctx.layer_transform_to_global(response.layer_id) {
+            widget_rect = to_global * widget_rect;
+        }
+        Self::Rect(widget_rect)
+    }
 }
 
-/// Show a tooltip at the current pointer position (if any).
-///
-/// Most of the time it is easier to use [`Response::on_hover_ui`].
-///
-/// See also [`show_tooltip_text`].
-///
-/// Returns `None` if the tooltip could not be placed.
-///
-/// ```
-/// # egui::__run_test_ui(|ui| {
-/// if ui.ui_contains_pointer() {
-///     egui::show_tooltip_at_pointer(ui.ctx(), ui.layer_id(), egui::Id::new("my_tooltip"), |ui| {
-///         ui.label("Helpful text");
-///     });
-/// }
-/// # });
-/// ```
-pub fn show_tooltip_at_pointer<R>(
-    ctx: &Context,
-    parent_layer: LayerId,
-    widget_id: Id,
-    add_contents: impl FnOnce(&mut Ui) -> R,
-) -> Option<R> {
-    ctx.input(|i| i.pointer.hover_pos()).map(|pointer_pos| {
-        let allow_placing_below = true;
-
-        // Add a small exclusion zone around the pointer to avoid tooltips
-        // covering what we're hovering over.
-        let mut pointer_rect = Rect::from_center_size(pointer_pos, Vec2::splat(24.0));
-
-        // Keep the left edge of the tooltip in line with the cursor:
-        pointer_rect.min.x = pointer_pos.x;
-
-        // Transform global coords to layer coords:
-        if let Some(from_global) = ctx.layer_transform_from_global(parent_layer) {
-            pointer_rect = from_global * pointer_rect;
+impl PopupAnchor {
+    /// The rect the popup should be shown at.
+    pub fn rect(self, popup_id: Id, ctx: &Context) -> Option<Rect> {
```

Comment:
I'm still not sure what this means. "Show popup in this rectangle" or "Show popup outside but next to this rectangle"?